x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/master/DedicatedMasterNodesDeciderService.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.autoscaling.master;
+
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingCapacity;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderContext;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderResult;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.StreamSupport;
+
+public class DedicatedMasterNodesDeciderService implements AutoscalingDeciderService {
+    public static final String NAME = "dedicated-masters-decider";
+
+    public static final int DEFAULT_NUMBER_OF_MASTER_NODES = 3;
+    public static final ByteSizeValue DEFAULT_MASTER_MEMORY = ByteSizeValue.ofGb(1);
+    public static final int DEFAULT_MIN_HOT_NODES = 6;
+    public static final ByteSizeValue DEFAULT_HOT_NODE_MIN_MEMORY = ByteSizeValue.ofGb(64);
+
+    public static final Setting<ByteSizeValue> MASTER_NODE_MEMORY = Setting.byteSizeSetting(
+        "master_node_memory",
+        (ignored) -> DEFAULT_MASTER_MEMORY.getStringRep(),
+        DEFAULT_MASTER_MEMORY,
+        ByteSizeValue.ofBytes(Long.MAX_VALUE)
+    );
+    public static final Setting<Integer> MIN_HOT_NODES = Setting.intSetting("min_hot_nodes", DEFAULT_MIN_HOT_NODES, 1);
+    public static final Setting<ByteSizeValue> MIN_HOT_NODE_MEMORY = Setting.byteSizeSetting(
+        "min_hot_node_memory",
+        (ignored) -> DEFAULT_HOT_NODE_MIN_MEMORY.getStringRep(),
+        ByteSizeValue.ofGb(1),
+        ByteSizeValue.ofBytes(Long.MAX_VALUE)
+    );
+
+    @Override
+    public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        int hotAndContentNodes = (int) StreamSupport.stream(context.state().nodes().spliterator(), false)
+            .filter(this::hasHotAndContentRole)
+            .count();
+        long totalHotAndContentNodesMemory = StreamSupport.stream(context.state().nodes().spliterator(), false)
+            .filter(this::hasHotAndContentRole)
+            .mapToLong(node -> getNodeMemory(node, context))
+            .sum();
```

Comment:
NIT: I believe it is possible to combine above into:

```
var summary = StreamSupport.stream(context.state().nodes().spliterator(), false)
            .filter(this::hasHotAndContentRole)
            .collect(summarizingLong(node -> getNodeMemory(node, context)))
summary.getCount()
summary.getSum()
```

Reply:
Nice, I didn't know about this. I'll update the PR. Thanks!